ddtrace/internal/encoding.py

```diff
@@ -34,17 +34,6 @@ def encode_traces(self, traces):
         normalized_traces = [[span.to_dict() for span in trace] for trace in traces]
         return self.encode(normalized_traces)
 
-    def encode_trace(self, trace):
```

Comment:
Why are we removing the `encode_trace` but not the `encode_traces` API from the Encoder classes here?

Reply:
The msgpack encoder uses a different base class that has a `put` method to put data in the buffer. `_EncoderBase` is the base of unbuffered encoders and they use the `encode_traces` API instead.

Reply:
Can we have the same API for buffered and non-buffered encoders? We could avoid any potential confusions by settling on a change to the API across encoders. We could replace `encode_trace` and `encode_traces` with `put`  (as you've done here) and include another function (`encode` in your new buffered encoder) that returns the object for the writer to act upon. At the moment, writers and encoders are still very much tied (e.g., the log writer only uses the non-buffered json encoder). But having a consistent API for all encoders would allow us to decouple it from the writer. Of course, this is beyond the scope of this refactor (which is good) but we are now in a place where the API is not consistent.

Reply:
Yeah I think we're realizing that encoding and writing are becoming more and more coupled (v0.5, partial flushing are examples of this). In the early days encoding was modular because we just naively encoded whatever trace data into json/msgpack but in order to get the perf benefits we're moving into custom formats.

My thinking here is that the encoding becomes an implementation detail of the writer and we should have different writers for each agent endpoint (WriterV04, WriterV05, etc).